akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala

```diff
@@ -335,13 +335,21 @@ object Actor {
    * State is updated by returning a new behavior that holds the new immutable
    * state.
    */
-  final case class Immutable[T](
+  final class Immutable[T] private (
     onMessage: (ActorContext[T], T) ⇒ Behavior[T],
-    onSignal:  (ActorContext[T], Signal) ⇒ Behavior[T] = Behavior.unhandledSignal.asInstanceOf[(ActorContext[T], Signal) ⇒ Behavior[T]])
+    onSignal:  PartialFunction[(ActorContext[T], Signal), Behavior[T]] = Behavior.unhandledSignal.asInstanceOf[PartialFunction[(ActorContext[T], Signal), Behavior[T]]])
     extends ExtensibleBehavior[T] {
     override def management(ctx: AC[T], msg: Signal): Behavior[T] = onSignal(ctx, msg)
     override def message(ctx: AC[T], msg: T) = onMessage(ctx, msg)
     override def toString = s"Immutable(${LineNumbers(onMessage)})"
+
+    def onSignal(onSignal: PartialFunction[(ActorContext[T], Signal), Behavior[T]]): Immutable[T] =
+      new Immutable(onMessage, onSignal)
+  }
```

Comment:
For the immutable see my comment here: https://github.com/akka/akka/pull/22798#discussion_r113715029

For the mutable we might want to do something when adding the builders, cc @raboof